Several SALChannelTest cases fail consistently on linux x86_64

### Expected behavior
All tests to pass on Fedora Core 33, as they did on Ubuntu 20.04 over the last several months.

### Actual behavior
Only the SALChannel tests fail, mostly with this error message `unexpected syscall whenReady(NIO.SelectorStrategy.now)`, which is some kind of race? I don't know if this is specific to Fedora, as I've since deleted my Ubuntu VPS. I see these errors occasionally on Android AArch64 too, which is why I then double-checked on linux, but they only randomly fail on Android when running the tests in `--parallel`, never serially. That flag makes no difference on Fedora, these tests always fail.

### Steps to reproduce
```
../swift-5.3.3-RELEASE-centos8/usr/bin/swift test -j 2 --enable-test-discovery --filter SALC
Test Suite 'Selected tests' started at 2021-02-13 06:59:37.867
Test Suite 'SALChannelTest' started at 2021-02-13 06:59:37.868
Test Case 'SALChannelTest.testBasicConnectWithClientBootstrap' started at 2021-02-13 06:59:37.868
/home/butta/swift-nio/Tests/NIOTests/SyscallAbstractionLayer.swift:506: error: SALChannelTest.testBasicConnectWithClientBootstrap : failed - unexpected syscall whenReady(NIO.SelectorStrategy.now)
/home/butta/swift-nio/Tests/NIOTests/SALChannelTests.swift:241: error: SALChannelTest.testBasicConnectWithClientBootstrap : failed - couldn't make a channel
Test Case 'SALChannelTest.testBasicConnectWithClientBootstrap' failed (10.003 seconds)
Test Case 'SALChannelTest.testBasicConnectedChannel' started at 2021-02-13 06:59:47.871
/home/butta/swift-nio/Tests/NIOTests/SyscallAbstractionLayer.swift:506: error: SALChannelTest.testBasicConnectedChannel : failed - unexpected syscall whenReady(NIO.SelectorStrategy.now)
<EXPR>:0: error: SALChannelTest.testBasicConnectedChannel : threw error "UnexpectedSyscall(syscall: NIOTests.UserToKernel.whenReady(NIO.SelectorStrategy.now))"
Test Case 'SALChannelTest.testBasicConnectedChannel' failed (10.001 seconds)
Test Case 'SALChannelTest.testBasicRead' started at 2021-02-13 06:59:57.872
/home/butta/swift-nio/Tests/NIOTests/SyscallAbstractionLayer.swift:506: error: SALChannelTest.testBasicRead : failed - unexpected syscall whenReady(NIO.SelectorStrategy.now)
/home/butta/swift-nio/Tests/NIOTests/SALChannelTests.swift:216: error: SALChannelTest.testBasicRead : XCTAssertNoThrow failed: threw error "UnexpectedSyscall(syscall: NIOTests.UserToKernel.whenReady(NIO.SelectorStrategy.now))" - 
/home/butta/swift-nio/Tests/NIOTests/SALChannelTests.swift:219: error: SALChannelTest.testBasicRead : failed - couldn't construct channel
Test Case 'SALChannelTest.testBasicRead' failed (10.001 seconds)
Test Case 'SALChannelTest.testClientBootstrapBindIsDoneAfterSocketOptions' started at 2021-02-13 07:00:07.873
/home/butta/swift-nio/Tests/NIOTests/SyscallAbstractionLayer.swift:506: error: SALChannelTest.testClientBootstrapBindIsDoneAfterSocketOptions : failed - unexpected syscall whenReady(NIO.SelectorStrategy.now)
/home/butta/swift-nio/Tests/NIOTests/SALChannelTests.swift:277: error: SALChannelTest.testClientBootstrapBindIsDoneAfterSocketOptions : failed - couldn't make a channel
Test Case 'SALChannelTest.testClientBootstrapBindIsDoneAfterSocketOptions' failed (10.001 seconds)
Test Case 'SALChannelTest.testWeSurviveIfIgnoringSIGPIPEFails' started at 2021-02-13 07:00:17.874
/home/butta/swift-nio/Tests/NIOTests/SyscallAbstractionLayer.swift:521: error: SALChannelTest.testWeSurviveIfIgnoringSIGPIPEFails : failed - unexpected syscall whenReady(NIO.SelectorStrategy.now)
/home/butta/swift-nio/Tests/NIOTests/SALChannelTests.swift:187: error: SALChannelTest.testWeSurviveIfIgnoringSIGPIPEFails : XCTAssertEqual failed: ("Optional(22)") is not equal to ("nil") - 
Test Case 'SALChannelTest.testWeSurviveIfIgnoringSIGPIPEFails' failed (10.001 seconds)
Test Case 'SALChannelTest.testWritesFromWritabilityNotificationsDoNotGetLostIfWePreviouslyWroteEverything' started at 2021-02-13 07:00:27.875
/home/butta/swift-nio/Tests/NIOTests/SyscallAbstractionLayer.swift:506: error: SALChannelTest.testWritesFromWritabilityNotificationsDoNotGetLostIfWePreviouslyWroteEverything : failed - unexpected syscall whenReady(NIO.SelectorStrategy.now)
/home/butta/swift-nio/Tests/NIOTests/SALChannelTests.swift:117: error: SALChannelTest.testWritesFromWritabilityNotificationsDoNotGetLostIfWePreviouslyWroteEverything : XCTAssertNoThrow failed: threw error "UnexpectedSyscall(syscall: NIOTests.UserToKernel.whenReady(NIO.SelectorStrategy.now))" - 
/home/butta/swift-nio/Tests/NIOTests/SALChannelTests.swift:120: error: SALChannelTest.testWritesFromWritabilityNotificationsDoNotGetLostIfWePreviouslyWroteEverything : failed - couldn't construct channel
Test Case 'SALChannelTest.testWritesFromWritabilityNotificationsDoNotGetLostIfWePreviouslyWroteEverything' failed (10.001 seconds)
Test Suite 'SALChannelTest' failed at 2021-02-13 07:00:37.876
	 Executed 6 tests, with 14 failures (1 unexpected) in 60.008 (60.008) seconds
Test Suite 'Selected tests' failed at 2021-02-13 07:00:37.876
	 Executed 6 tests, with 14 failures (1 unexpected) in 60.008 (60.008) seconds
```

### SwiftNIO version/commit hash

latest HEAD - ae3ea78

### Swift & OS version
```
Swift version 5.3.3 (swift-5.3.3-RELEASE)
Target: x86_64-unknown-linux-gnu
```
`uname -r` is `5.10.9-201.fc33.x86_64`.

--- a/Tests/NIOTests/SyscallAbstractionLayer.swift
+++ b/Tests/NIOTests/SyscallAbstractionLayer.swift
@@ -118,7 +118,7 @@ extension LockedBox where T == UserToKernel {
     func assertParkedRightNow(file: StaticString = #file, line: UInt = #line) throws {
         SAL.printIfDebug("\(#function)")
         let syscall = try self.waitForValue()
-        if case .whenReady = syscall {
+        if case .whenReady(.block) = syscall {
             return
         } else {
             XCTFail("unexpected syscall \(syscall)", file: (file), line: line)
@@ -421,6 +421,12 @@ extension EventLoop {
                             _ body: @escaping () throws -> T) throws -> T {
         let hookedSelector = ((self as! SelectableEventLoop)._selector as! HookedSelector)
         let box = LockedBox<Result<T, Error>>()
+
+        // To prevent races between the test driver thread (this thread) and the EventLoop (another thread), we need
+        // to wait for the EventLoop to finish its tick and park itself. That makes sure both threads are synchronised
+        // so we know exactly what the EventLoop thread is currently up to (nothing at all, waiting for a wakeup).
+        try hookedSelector.userToKernel.assertParkedRightNow()
+
         self.execute {
             do {
                 try box.waitForEmptyAndSet(.init(catching: body))
@@ -430,7 +436,9 @@ extension EventLoop {
         }
         try hookedSelector.assertWakeup(file: (file), line: line)
         try syscallAssertions()
-        try hookedSelector.userToKernel.assertParkedRightNow() // We need the EventLoop to finish running its tick.
+
+        // Here as well, we need to synchronise and wait for the EventLoop to finish running its tick.
+        try hookedSelector.userToKernel.assertParkedRightNow()
         return try box.takeValue().get()
     }
 }
@@ -441,22 +449,15 @@ extension EventLoopFuture {
     /// Using a plain `EventLoopFuture.wait()` together with the SAL would require you to spin the `EventLoop` manually
     /// which is error prone and hard.
     internal func salWait() throws -> Value {
-        assert(Thread.isMainThread)
         let box = LockedBox<Result<Value, Error>>()
-        let hookedSelector = ((self.eventLoop as! SelectableEventLoop)._selector as! HookedSelector)
 
-        self.eventLoop.execute {
-            self.whenComplete { result in
-                do {
-                    try box.waitForEmptyAndSet(result)
-                } catch {
-                    box.value = .failure(error)
-                }
+        XCTAssertNoThrow(try self.eventLoop.runSAL() {
+            self.whenComplete { value in
+                // We can bang this because the LockedBox is empty so it'll immediately succeed.
+                try! box.waitForEmptyAndSet(value)
             }
-        }
-        try hookedSelector.assertWakeup()
-        try hookedSelector.userToKernel.assertParkedRightNow() // We need the EventLoop to finish running its tick.
-        return try box.takeValue().get()
+        })
+        return try box.waitForValue().get()
     }
 }
 

